### PR TITLE
Perception: Separate lane line model and new post processing

### PR DESCRIPTION
**Pull request by xiaoshuliu**
Trained a new model to predict lanes with different lane labels separately.
Implemented new post processing algorithm. Now it is much simpler and faster taking advantage of the new model. 
Currently all the new features are not turned on. I will send another commit (modify some parameters) to turn on it as soon as the new model is written into the docker image.

### CODE REVIEW

File: modules/perception/obstacle/camera/detector/yolo_camera_detector/yolo_camera_detector.cc [hunk 1]
```diff
@@ -393,17 +389,15 @@ bool YoloCameraDetector::Lanetask(const cv::Mat &frame, cv::Mat *mask) {
...
-  cv::Mat tmp(lane_output_height_lane_, lane_output_width_lane_, CV_32FC1);
-  memcpy(tmp.data, seg_blob->cpu_data() +
-        lane_output_width_lane_ * lane_output_height_lane_,
-         lane_output_width_lane_ * lane_output_height_lane_ * sizeof(float));
+  for (int i = 0; i < 13; ++i) {
+    cv::Mat tmp(lane_output_height_lane_, lane_output_width_lane_, CV_32FC1);
+    memcpy(tmp.data, seg_blob->cpu_data() + lane_output_width_lane_ *
+          lane_output_height_lane_ * i,
+          lane_output_width_lane_ * lane_output_height_lane_ * sizeof(float));
+    cv::resize(tmp, tmp, cv::Size(960, 384), 0, 0);
```
Comment (techoe) [inline modules/perception/obstacle/camera/detector/yolo_camera_detector/yolo_camera_detector.cc]: Please change 960 & 384 to const values

File: modules/perception/obstacle/camera/lane_post_process/cc_lane_post_processor/cc_lane_post_processor.h [hunk 2]
```diff
@@ -130,7 +130,16 @@ class CCLanePostProcessor : public BaseCameraLanePostProcessor {
...
   int image_height_ = 1920;
+
   cv::Rect roi_;
+  int roi_height = 768;
+  int roi_start = 312;
+  int roi_width = 1920;
+
+  Eigen::Matrix<double, 3, 3> trans_mat_;
+  Eigen::Matrix<double, 3, 3> trans_mat_inv;
+  std::vector<std::vector<Eigen::Matrix<double, 2, 1>>> xy_points;
```
Comment (techoe) [inline modules/perception/obstacle/camera/lane_post_process/cc_lane_post_processor/cc_lane_post_processor.h]: You don't have to change now but later please change all std::vector to fixed sized vector (e.g. fixed_vector or circular_buffer )

File: modules/perception/obstacle/camera/lane_post_process/common/type.h [hunk 3]
```diff
@@ -82,6 +82,7 @@ const cv::Scalar kGreenYellow(47, 255, 173);
 
 // delay time for visualization
 constexpr int kDelayTime = 0;
+const size_t minNumPoints = 5;
```
Comment (techoe) [inline modules/perception/obstacle/camera/lane_post_process/common/type.h]: Please add comment why the min Num point is 5

File: modules/perception/obstacle/camera/visualizer/glfw_fusion_viewer.cc [hunk 4]
```diff
@@ -63,6 +63,13 @@ std::vector<std::vector<int>> GLFWFusionViewer::s_color_table = {
     std::vector<int>{255, 0, 255}, std::vector<int>{255, 255, 0},
     std::vector<int>{255, 128, 0}};
 
+std::vector<cv::Scalar> lane_map_colors = {
+    cv::Scalar(0, 0, 0), cv::Scalar(100, 240, 0), cv::Scalar(150, 180, 0),
```
Comment (techoe) [inline modules/perception/obstacle/camera/visualizer/glfw_fusion_viewer.cc]: Please make them to have one color in one line and add color name (at least something similar)

Comment (techoe) [reply modules/perception/obstacle/camera/visualizer/glfw_fusion_viewer.cc]: Semantic annotation (L0, R0, ... ) will be also useful

File: modules/perception/obstacle/camera/visualizer/glfw_fusion_viewer.cc [hunk 5]
```diff
@@ -1248,20 +1261,23 @@ bool GLFWFusionViewer::draw_lane_objects_image(cv::Mat* image_mat) {
...
-  cv::Scalar lane_map_color(0, 255, 255);  // yellow for lane line mark
+  // cv::Scalar lane_map_color(0, 255, 255);  // yellow for lane line mark
   int x_offset = 0;
   int y_offset = lane_start_y_pos_;
   int x0 = x_offset;
   int y0 = y_offset;
   int x1 = image_mat->cols - 1;
   int y1 = image_mat->rows - 1;
-
+  lane_map_threshold_ = 0.5;
```
Comment (techoe) [inline modules/perception/obstacle/camera/visualizer/glfw_fusion_viewer.cc]: 0.5 as an const in header file

File: modules/perception/obstacle/camera/visualizer/glfw_fusion_viewer.cc [hunk 6]
```diff
@@ -1310,30 +1326,81 @@ bool GLFWFusionViewer::draw_lane_objects_image(cv::Mat* image_mat) {
...
+      float d = lane_objects_->at(k).img_curve.d;
+
+      for (float l = start; l <= end; l++) {
+        cv::circle(*image_mat,
+                   cv::Point(static_cast<int>(GetPolyValue(a, b, c, d, l)),
+                             static_cast<int>(l)),
+                             2, lane_object_color, -1);
+      }
+    } else {
+      CalibrationConfigManager* calibration_config_manager =
```
Comment (techoe) [inline modules/perception/obstacle/camera/visualizer/glfw_fusion_viewer.cc]: Please get this homography at render() function and save globally (e.g. H_car2camera_ ) and use here. Other's may use the homography other than lane line for each frame.

File: modules/perception/obstacle/onboard/camera_process_subnode.cc [hunk 7]
```diff
@@ -124,11 +124,32 @@ void CameraProcessSubnode::ImgCallback(const sensor_msgs::Image &message) {
...
+  cv::Mat mask_color(mask.rows, mask.cols, CV_32FC1);
   if (FLAGS_use_whole_lane_line) {
-    cv::Mat mask1;
-    detector_->Lanetask(img, &mask1);
-    mask += mask1;
+    std::vector<cv::Mat> masks;
+    detector_->Lanetask(img, &masks);
+    mask_color.setTo(cv::Scalar(0));
+    ln_msk_threshold_ = 0.9;
+    for (int c = 0; c < 13; ++c) {
```
Comment (techoe) [inline modules/perception/obstacle/onboard/camera_process_subnode.cc]: 13

File: modules/perception/obstacle/onboard/camera_process_subnode.h [hunk 8]
```diff
@@ -119,7 +119,7 @@ class CameraProcessSubnode : public Subnode {
   bool pb_obj_ = false;  // Objects
   apollo::canbus::Chassis chassis_;
   bool pb_ln_msk_ = false;  // Lane marking mask
-  const float ln_msk_threshold_ = 0.95f;
+  float ln_msk_threshold_ = 0.95f;
```
Comment (techoe) [inline modules/perception/obstacle/onboard/camera_process_subnode.h]: Why it is not const?

Comment (xiaoshuliu) [reply modules/perception/obstacle/onboard/camera_process_subnode.h]: It is different in old model and new model

**Pull request merged**